File: Python/1190.reverse-substrings-between-each-pair-of-parentheses.py

class Solution(object):
    def reverseParentheses(self, s):
        """
        :type s: str
        :rtype: str
        :desc: 第二次根据括号考虑栈，通过
        """
        s = s.replace('()', '')
        groups = {}
        group_index = 0
        brackets = []
        prev = '('
        for idx, c in enumerate(s):
            if c == '(':
                brackets.append(idx)
                if prev == ')':
                    group_index += 1
                    prev = '('
            if c == ')':
                prev = ')'
                left_idx = brackets.pop()
                if group_index not in groups:
                    groups[group_index] = {}
                groups[group_index][left_idx] = idx

        for group in groups.values():
            for start in sorted(group, reverse=True):
                end = group[start]
                start += 1
                substr = s[start:end]
                s = s[:start] + substr[::-1] + s[end:]
        return s.replace('(', '').replace(')', '')

    def reverseParentheses_first(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        :type s: str
        :rtype: str
        :desc: 第一次虽然考虑到用栈，但是没有考虑全面，通过case：19/37.
        """
        arr = []
        sarr = s.split('(')
        for i in sarr[:-1]:
            arr.append(i)
        for i in sarr[len(sarr) - 1].split(')'):
            arr.append(i)
        lst = []
        total = len(arr)
        mid = (total+1)/2
        for i in range(total):
            if i % 2 == 0:
                lst.append(arr[i])
            else:
                if i < mid:
                    temp = arr[total - 1 - i]
                    arr[total - 1 - i] = arr[i]
                    arr[i] = temp
                lst.append(arr[i][::-1])

        return "".join(lst)
        pass

File: PHP/1190.reverse-substrings-between-each-pair-of-parentheses.php
<?php

class Solution {

    /**
     * @desc 参考下其他人的解题思路。核心思想就是栈操作，遇到左括号栈顶就压入空串，遇到右括号就反转栈顶并与栈顶第二个元素合并，其他情况栈顶直接累加元素，最后输出栈内唯一元素。
     * @param String $s
     * @return String
     */
    function reverseParentheses($s) {
        echo $s."\n";
        $arr = [""];
        for($i=0; $i<strlen($s); $i++) {
            $c = $s[$i];
            $idx = count($arr);
            if ($c == '(') {
                array_push($arr, "");
                echo $c."\t".implode(',', $arr)."\n";
            } else if ($c == ')') {
                //每次遇到右括号，则相邻的2个进行合并
                $arr[$idx-2] = $arr[$idx-2].strrev($arr[$idx - 1]);
                echo $c."\t".implode(',', $arr)."\n";
                array_pop($arr);
            } else {
                $arr[$idx-1] .= $c;
                echo $c."\t".implode(',', $arr)."\n";
            }
            
        }
        return implode('', $arr);
    }
}

$s1 = "(ed(et(oc))el)";
$s2 = "a(bcdefghijkl(mno)p)q";
$s = new Solution();
$s->reverseParentheses($s1);
$s->reverseParentheses($s2);
